File: repos/12865/12865/12865.cpp
#include<iostream>
using namespace std;

int knapsack[100001][100];
int N, K;
int V, W;

int main() {
	cin >> N >> K;
	cin >> W >> V;
	for (int j = W; j <= K; j++) {
		knapsack[j][0] = V;
	}
	for (int i = 1; i < N; i++) {
		cin >> W >> V;
		
		for (int j = 0; j < W; j++) {
			knapsack[j][i] = knapsack[j][i-1];
		}
		for (int j = W; j <= K; j++) {
			knapsack[j][i] = max(knapsack[j][i - 1], knapsack[j - W][i - 1] + V);
		}
	}

	cout << knapsack[K][N-1];
	
}

File: week_4/12865.py
import sys
import copy

N,K = map(int, sys.stdin.readline().split())

knapsack = [[0 for _ in range(K+1)] for _ in range(N+1)]
products = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]

for i in range(1,N+1):
    weight , value = products[i-1]
    if weight <= K:
        for j in range(1, weight):
            knapsack[i][j] = knapsack[i-1][j]
        for j in range(weight, K+1):
            knapsack[i][j] = max(knapsack[i-1][j-weight] + value , knapsack[i-1][j])
    else:
        knapsack[i] = knapsack[i-1]

print(knapsack[N][K])
